### PR TITLE
Index series, not chunks

### LINKED ISSUE #433: Chunk Store should index series, not chunks
Discussed with @aaron7, our index should be over series, not chunks.  This will enable lazily loading of chunk data in the future.

Indexes should be:
- userid:day -> series id -> label set
- userid:day:metric_name -> series id -> label set
- userid:day:metric_name:label_name -> #(label_value) : series_id -> (label_value, label_set)
- userid:day:series_id -> chunk_id -> chunk

### LINKED ISSUE #884: Don't query very high cardinality labels
(Building on an idea from #209)

We have a common class of query on very-high cardinality metrics (~800k timeseries) which actually only need to read a small number (~500) of those timeseries.  These queries usually involve multiple (4) matchers, 1-2 of them match pretty much all timeseries for that metric (like the job matcher), and 1-2 end up filtering it down to the 500 series in the end.

With the previous chunk store, this would pretty much always OOM, as it ended up reading millions of chunk descriptors before intersecting them away.  With the series store (#875), these queries no longer OOM, they're just slow, measures in 10s of seconds.

I propose we keep a in-querier-process cache of “matcher effectiveness”, measured as the number of series matched by a matcher that making it into a final result.  The idea would be that this would allow us to detect and ignore matchers which match lots of series that don’t survive the intersection.

Or perhaps alternatively we could keep a cache of observed label cardinality and just not read labels beyond a certain size (ignoring those matchers).  Eventually if we are forced to ignore all the matchers, we could fail the whole query with a cardinality error.

TLDR: both these proposal involve ignoring matchers and potentially reading more chunks, but avoiding doing very high cardinality queries on the index, saving memory and hopefully decreasing latency.

WDYT?

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #433 , fixes #884 

Consist of 4 changes:

---

**Move compositeSchema abstraction up to componsiteStore.**

Promote the composite schema abstraction to "composite chunk store" - a chunk store which delegates to different chunk stores based on time.  This allows us to vary the store implementation over time, and not just the schema.  This will unblock the new bigtable storage adapter (using columns instead of rows), and allow us to more easily implement the iterative intersections and indexing of series instead of chunks.

Corner case when a writing chunks which span multiple stores: they are written to both stores, and instead of using the chunk start/end we use the schema start/end.  This will lead to duplication of the index entries on schema migrations, but is actually already the case for day boundaries anyway.  It will lead to duplicate writes of the chunk on schema migrations - they should be deduped by the underlying store.

---

**Index series, not chunks**

We should index the series, not chunks; this will reduce the number of entries in the index by `replication factor * (chunk size / bucket size)`, or 3 * 6hrs / 24hrs - ie 12x. This will however mean we need another index from series to chunks, introducing 1 extra write an N extra reads per query.  Expectation is a reduction in query latency (and bigtable query usage, and memory cost) by 12x, and then an increase by 2x as we have to do a bunch of queries.

This change introduces the seriesStore, a new chunk store implementation that, combined with the v9 schema, indexes series not chunks.

I tried to adapt the original chunk store to support this style of indexing - easy on the write path, but the read path became even more of a rats nest.  So I factored out the common bits as best I could and made a new chunk store.

---

**Add heapCache, a cache that uses a heap for evictions.**

heapCache is a simple string -> interface{} cache which uses a heap to manage evictions.  O(log N) inserts and updates, O(1) gets.

---

**Skip index queries for high cardinality labels.**

Firstly, cache the length of index rows we query (by the hash and range key).  Secondly, fail for rows with > 100k, either because the cache told us so, or because we read them.  Finally, allow matchers to fail on cardinality errors but proceed with the query (as long as at least one matcher succeeds), and then filter results.

Notably, after this change, queries on two high-cardinality labels that would have results in a small number of series will fail.

### CODE REVIEW

**Comment by tomwilkie** [comment]
This will fail when queries cross schema buckets; plan is to move the `compositeSchema` code up one level to have a `compositeStore`, which multiplexes queries to different chunk stores based on time.

**Comment by tomwilkie** [comment]
Got this deployed in our dev env and it seems to work.  Hard to draw any conclusions as there aren't any taxing queries, but there seems to be a drop in bigtable latency in line with a reduction the amount of data we're reading.

**Comment by tomwilkie** [comment]
Running in prod for a few days now and seen the worse queries go from minutes to seconds, in line with a 6-10x latency improvement.  Once we rotate out the index tables, will report on their size.

**Comment by tomwilkie** [comment]
The was a bug in the way we handled chunk spanning the transition; fixed now.

File: cmd/ingester/main.go [hunk 1]
```diff
@@ -70,7 +70,7 @@ func main() {
 		os.Exit(1)
 	}
 
-	chunkStore, err := chunk.NewStore(chunkStoreConfig, schemaConfig, storageClient)
+	chunkStore, err := chunk.NewCompositeStore(chunkStoreConfig, schemaConfig, storageClient)
```
Comment (bboreham) [inline cmd/ingester/main.go]: ISTM that you could have left the exported name as NewStore() to minimise impact elsewhere.

File: pkg/chunk/chunk_store.go [hunk 2]
```diff
@@ -93,22 +93,34 @@ func (c *store) Stop() {
 
 // Put implements ChunkStore
 func (c *store) Put(ctx context.Context, chunks []Chunk) error {
+	for _, chunk := range chunks {
+		if err := c.PutOne(ctx, chunk.From, chunk.Through, chunk); err != nil {
```
Comment (bboreham) [inline pkg/chunk/chunk_store.go]: This seems like rather a large change in performance characteristics, if you have a number of chunks to write.

File: pkg/chunk/chunk_store_test.go [hunk 3]
```diff
@@ -14,22 +14,45 @@ import (
...
 	"github.com/weaveworks/cortex/pkg/util/extract"
 	"golang.org/x/net/context"
 
 	"github.com/weaveworks/common/test"
 	"github.com/weaveworks/common/user"
 )
 
+var schemas = []struct {
+	name              string
+	fn                func(cfg SchemaConfig) Schema
```
Comment (bboreham) [inline pkg/chunk/chunk_store_test.go]: I was expecting `storeFn` to come in this commit, even though they would all be the same at this point.

**Comment by cboggs** [comment]
Seems to me that this schema should also allow (or lay the groundwork for) successful queries without invariant metric names... am I thinking on the right path there?

File: cmd/ingester/main.go (hunk 1 above)
Comment (tomwilkie) [reply cmd/ingester/main.go]: Good idea, done.

**Comment by tomwilkie** [comment]
>Seems to me that this schema should also allow (or lay the groundwork for) successful queries without invariant metric names... am I thinking on the right path there?

It could, but still doesn't avoid the problem of hotspotting the "__name__" row.  Although it would reduce the load on that row by 12x, so might be doable at this point.

**Comment by tomwilkie** [comment]
> This seems like rather a large change in performance characteristics, if you have a number of chunks to write.

Yes, it potentially is.  I don't think it will effect normal operation, as I think we only flush one chunk at once normally - we've been running this for a week or so, haven't notices anything.  But may effect shutdown flushes.

I think the correct solution is to potentially split out the `ChunkStore` (get and put chunks by ID) and `ChunkIndex` (write entries and find entries by matchers), something that is starting to happen with the `seriesStore` anyway.  Then the composite store can write the index entries and the chunks once, in a batch.  WDYT?

**Comment by tomwilkie** [comment]
> Do we really need to write another cache?

How many do we have?  I see the chunk caches (which are fundamentally different things), a vendored prometheus treecache (something to do with zookeeper) and some caches in k8s/client-go.  I guess the grpc connection pool is a cache, but thats a different beast too.  We also cache a single result in the chunk iterator, but thats just a single result.

I'm not actually aware of any other caches (like this) in use in the codebase.   We could use github.com/bluele/gcache, but I was never a fan of that library.

OTOH, not clear using a heap for evictions if the best idea; we could thread a link list through the entries and reorder that to get LRU.

**Comment by tomwilkie** [comment]
I've tidied up this PR so it should be more reviewable.  I see two things left todo: figure out what do to with chunks on store boundaries (try and bring back the batching) and accurately record cardinality or rows for multi-day queries.

Let me know if you get a chance to take a look @bboreham @cboggs @gouthamve.

**Comment by bboreham** [comment]
By "we" I meant the Go community.

I've used `github.com/patrickmn/go-cache` successfully in https://github.com/weaveworks/scope, to replace `github.com/bluele/gcache` for performance reasons.

**Comment by bboreham** [comment]
> I think we only flush one chunk at once normally

If the flush queue gets very large (and we've had it in the millions many times) this assumption breaks down.

> Then the composite store can write the index entries and the chunks once, in a batch. WDYT?

I wrote some thoughts at https://github.com/weaveworks/cortex/issues/684#issuecomment-408143379

File: pkg/chunk/schema_config.go [hunk 4]
```diff
@@ -59,6 +60,7 @@ func (cfg *SchemaConfig) RegisterFlags(f *flag.FlagSet) {
 	f.Var(&cfg.V6SchemaFrom, "dynamodb.v6-schema-from", "The date (in the format YYYY-MM-DD) after which we enable v6 schema.")
 	f.Var(&cfg.V7SchemaFrom, "dynamodb.v7-schema-from", "The date (in the format YYYY-MM-DD) after which we enable v7 schema (Deprecated).")
 	f.Var(&cfg.V8SchemaFrom, "dynamodb.v8-schema-from", "The date (in the format YYYY-MM-DD) after which we enable v8 schema (Deprecated).")
+	f.Var(&cfg.V9SchemaFrom, "dynamodb.v9-schema-from", "The data (in the format YYYY-MM-DD) after which we enable v9 schema (Series indexing).")
```
Comment (csmarchbanks) [inline pkg/chunk/schema_config.go]: typo - data should be date

File: pkg/chunk/schema_config.go [hunk 5]
```diff
@@ -60,6 +60,7 @@ func (cfg *SchemaConfig) RegisterFlags(f *flag.FlagSet) {
...
```
Comment (csmarchbanks) [inline pkg/chunk/schema_config.go]: typo, data should be date

File: pkg/chunk/series_store.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+// Get implements Store
+func (c *seriesStore) Get(ctx context.Context, from, through model.Time, allMatchers ...*labels.Matcher) ([]Chunk, error) {
+	log, ctx := newSpanLogger(ctx, "ChunkStore.Get")
+	defer log.Span.Finish()
+	level.Debug(log).Log("from", from, "through", through, "matchers", len(allMatchers))
+
+	// Validate the query is within reasonable bounds.
+	shortcut, err := c.validateQuery(ctx, from, &through)
```
Comment (csmarchbanks) [inline pkg/chunk/series_store.go]: Why is through a pointer?

File: pkg/chunk/series_store.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+	// Lookup the series in the index to get the chunks.
+	chunkIDs, err := c.lookupChunksBySeries(ctx, from, through, seriesIDs)
+	if err != nil {
+		return nil, err
+	}
+	level.Debug(log).Log("Chunk IDs", len(chunkIDs))
+
+	// Protect ourselves against OOMing.
+	if len(chunkIDs) > c.cfg.QueryChunkLimit {
```
Comment (csmarchbanks) [inline pkg/chunk/series_store.go]: Would it make more sense to do this after filtering out the chunks by time?

**Comment by csmarchbanks** [comment]
I also have some concerns about putting one chunk vs many chunks, and added a comment with a possible idea. I like the idea of splitting `ChunkStore` and `ChunkIndex`, or some of Bryan's comments, but perhaps not as part of this PR

**Comment by tomwilkie** [comment]
> I've used github.com/patrickmn/go-cache successfully in https://github.com/weaveworks/scope, to replace github.com/bluele/gcache for performance reasons.

I've looked at go-cache, it still uses a background goroutine to periodically expunge entries from the cache.  AFAICT this means the cache can grow without bounds between these periods, and it also locks the entire cache to do this.  The heap cache isn't ideal, I'm going to update it to use a simple FIFO list for evictions, but I think its better that go-cache.

**Pull request merged**